### PR TITLE
feature/TIN-19

### PR DESCRIPTION
**Pull request by tsenovi**
image

### CODE REVIEW

**Comment by todornikolov1** [comment]
Please, in the future use branch's name as the card number of the task in your Trello board. For this tasks should be TIN-19. You can keep the future/ (or fix/) prefix if you want

File: src/main/java/com/volasoftware/tinder/constants/Constants.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.volasoftware.tinder.constants;
+
+public class Constants {
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/constants/Constants.java]: It's a good idea to keep constants in separate class. As project growing, the number of constants will increase, so think to put and separate them in different classes according the context.

File: src/main/java/com/volasoftware/tinder/dtos/AccountDto.java [hunk 2]
```diff
@@ -6,16 +6,14 @@
...
+import lombok.*;
 
-@Data
+@Getter
+@Setter
 @AllArgsConstructor
 @NoArgsConstructor
 @Builder
-@ApiModel(description = "Public account information")
+@ApiModel(value = "Public model", description = "Public account information")
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/dtos/AccountDto.java]: value = "Public model" ?

File: src/main/java/com/volasoftware/tinder/services/contracts/AuthenticationService.java [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
-  List<AccountDto> getAll();
+    List<AccountDto> getAll();
+
+    AccountDto register(RegisterRequest registerRequest);
 
-  AccountDto register(RegisterRequest registerRequest);
+    AccountDto getAccountByEmail(String email);
 
-  Account getAccountByEmail(String email);
+    String verify(String token);
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/contracts/AuthenticationService.java]: verify() returns a String? It's not quite intuitive for me

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -67,23 +68,37 @@ public AccountDto register(RegisterRequest registerRequest) {
...
     @Override
-    public Account getAccountByEmail(String email) {
-        return accountRepository.findOneByEmail(email)
-                .orElseThrow(() -> new AccountNotFoundException("Account not found!"));
+    public AccountDto getAccountByEmail(String email) {
+        return AccountMapper.INSTANCE.accountToAccountDto(
+                accountRepository.findOneByEmail(email)
+                        .orElseThrow(
+                                () -> new AccountNotFoundException(Constants.ACCOUNT_NOT_FOUND)
+                        )
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java]: Nested like that is more difficult to read and understand, for me at least :) 

You can assign it to Account variable and then pass that new variable to AccountMapper.INSTANCE.accountToAccountDto( ... ).

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 5]
```diff
@@ -43,7 +43,7 @@ public void send(String receiverEmail, String receiverFirstName, String token) {
     }
 
     private String buildEmail(String name, String token) {
-        String link = "http://localhost:8080/api/v1/users/register/confirm?token=" + token;
+        String link = "http://localhost:8080/api/v1/users/verify?token=" + token;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: `http://localhost:8080` is environment dependent. Should be move as value in application.properties

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java [hunk 6]
```diff
@@ -25,13 +31,12 @@ public class VerificationTokenServiceImpl implements VerificationTokenService {
...
     @Override
     public void generateToken(Account account) {
-
         String uuidToken = UUID.randomUUID().toString();
         VerificationToken verificationToken = new VerificationToken(
                 uuidToken,
                 LocalDateTime.now(),
                 LocalDateTime.now(),
-                LocalDateTime.now().plus(2, ChronoUnit.DAYS),
+                LocalDateTime.now().plusDays(2),
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: Please, don't use hardcoded value

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java [hunk 7]
```diff
@@ -40,4 +45,38 @@ public void generateToken(Account account) {
...
+        isTokenExist(optionalVerificationToken);
+
+        VerificationToken verificationToken = optionalVerificationToken.get();
+        isEmailAlreadyVerified(verificationToken);
+
+        LocalDateTime expiresAt = verificationToken.getExpiresAt();
+        isTokenExpired(expiresAt);
+
+        verificationToken.setVerifiedAt(LocalDateTime.now());
+        verificationTokenRepository.save(verificationToken);
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: Almost in every case use saved data (returned from .save()) afterwards.

File: src/test/java/com/volasoftware/tinder/controllers/AuthenticationControllerTest.java [hunk 8]
```diff
@@ -36,6 +39,24 @@ public void setUp(WebApplicationContext webApplicationContext) {
...
     }
 
+    @Test
+    void givenVerificationTokenWhenEmailIsNotVerifiedThenSuccessfulVerificationMsg() throws Exception {
+
+        //Given
+        String token = "73275233-bf2a-4c60-b2c9-10d4d02d8160";
+
+        // When
+        given(authenticationService.verify(any(String.class))).willReturn(Constants.VERIFIED);
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/controllers/AuthenticationControllerTest.java]: Is 73275233-bf2a-4c60-b2c9-10d4d02d8160 equal to passed argument in authenticationService.verify(any(String.class), if yes, you can use it

File: src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    private static final String PASSWORD = "password";
+
+    @Autowired
+    VerificationTokenRepository verificationTokenRepository;
+
+    @Autowired
+    AccountRepository accountRepository;
+
+    @AfterEach
+    void tearDown() {
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java:34]: interesting name :)

File: src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        verificationTokenRepository.deleteAll();
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java:35]: I hope you have a separate instance of the database for the tests

**Review by todornikolov1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/volasoftware/tinder/dtos/AccountDto.java (hunk 2 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/dtos/AccountDto.java]: Meant to be "Public account"

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java (hunk 4 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java]: Fixed

File: src/main/java/com/volasoftware/tinder/services/contracts/AuthenticationService.java (hunk 3 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/contracts/AuthenticationService.java]: Fixed. Now the method returns the public account information

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java (hunk 7 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: Now I`m using the account from the token as a return value for the method.

File: src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java (hunk 10 above)
Comment (tsenovi) [reply src/test/java/com/volasoftware/tinder/repositories/VerificationTokenRepositoryTest.java:35]: All tests are executed against in-memory database - H2 in separate application.properties file.

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java (hunk 6 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: Refactored code by setting new Constant in separate class

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 5 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: EmailService structure was refactored

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java [hunk 11]
```diff
@@ -35,17 +35,23 @@ public class AuthenticationServiceImpl implements AuthenticationService {
     @Autowired
     private VerificationTokenService verificationTokenService;
 
+    @Autowired
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java]: Read this article: https://odrotbohm.de/2013/11/why-field-injection-is-evil/

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 12]
```diff
@@ -42,73 +60,27 @@ public void send(String receiverEmail, String receiverFirstName, String token) {
...
-                "</div></div>";
+    private String buildVerificationMail(String name, String token) {
+        String link = baseUrl + verifyUrl + token;
+
+        byte[] contentBytes = readFile(MailConstant.VERIFICATION_FILE);
+
+        String htmlContent = new String(contentBytes);
+        String htmlWithLink = htmlContent.replace(MailConstant.ACTIVATION_LINK, link);
+
+        return htmlWithLink;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]:
```suggestion
        return htmlContent.replace(MailConstant.ACTIVATION_LINK, link);
```

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java [hunk 13]
```diff
@@ -20,24 +25,52 @@ public class VerificationTokenServiceImpl implements VerificationTokenService {
...
 
-        verificationTokenRepository.save(verificationToken);
+        return verificationTokenRepository.save(verificationToken);
+    }
+
+    @Override
+    public Account verifyToken(String token) {
+
+        Optional<VerificationToken> optionalVerificationToken = verificationTokenRepository.findByToken(token);
+        isTokenExist(optionalVerificationToken);
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: if token isn't found just throw VerificationTokenExpiredException (see .orElseThrow() over optional)

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java (hunk 11 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java]: Constructor injection was implemented where needed.

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 12 above)
Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: Is the suggestion applied? I don't see changes

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java (hunk 13 above)
Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: Any changes here?

File: src/test/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImplTest.java [hunk 14]
```diff
@@ -144,4 +163,31 @@ private RegisterRequest getRegisterRequest(String firstName, String lastName, St
 
         return registerRequest;
     }
+
+    private VerificationToken generateVerificationToken(
+            Account account) {
+
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImplTest.java]: You can remove the line

File: src/test/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        //then
+        verify(repository, times(1)).save(verificationToken);
+
+        assertNotNull(expectedAccount);
+        assertEquals(account.getEmail(), expectedAccount.getEmail());
+    }
+
+    private VerificationToken generateVerificationToken(
+            Account account) {
+
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImplTest.java]: remove the line

File: src/test/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        VerificationToken verificationToken = new VerificationToken();
+        String randomToken = UUID.randomUUID().toString();
+        verificationToken.setToken(randomToken);
+        verificationToken.setAccount(account);
+        verificationToken.setCreatedDate(LocalDateTime.now());
```
Comment (todornikolov1) [inline src/test/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImplTest.java]: In tests, it's not good idea to use LocalDateTime.now(), its hard to be tested, just create concrete one instead (ex. LocalDateTime.of(...))

**Review by todornikolov1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java (hunk 13 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java]: https://github.com/tsenovi/spring-tinder-application/pull/16/commits/4a8384483b194ca45f25d61fe18718af4767ce95

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 12 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: https://github.com/tsenovi/spring-tinder-application/pull/16/commits/057bf99c4d739e7f44606a87a63a146c45412f09

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java [hunk 17]
```diff
@@ -26,26 +26,25 @@
...
-    private JwtService jwtService;
+    private final JwtService jwtService;
 
-    @Autowired
-    private VerificationTokenService verificationTokenService;
+    private final VerificationTokenService verificationTokenService;
 
-    @Autowired
-    private EmailValidator emailValidator;
+    private final EmailService emailService;
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java:33]: Do we use it?

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 18]
```diff
@@ -42,73 +59,26 @@ public void send(String receiverEmail, String receiverFirstName, String token) {
...
-                "      </td>\n" +
-                "      <td width=\"10\" valign=\"middle\"><br></td>\n" +
-                "    </tr>\n" +
-                "    <tr>\n" +
-                "      <td height=\"30\"><br></td>\n" +
-                "    </tr>\n" +
-                "  </tbody></table><div class=\"yj6qo\"></div><div class=\"adL\">\n" +
-                "\n" +
-                "</div></div>";
+    private String buildVerificationMail(String name, String token) {
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: This method is not related with EmailService. Just create the content before calling emailService.send() passing as an argument in it

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 19]
```diff
@@ -19,14 +27,23 @@ public class EmailServiceImpl implements EmailService {
...
+    @Value("${server.verify-url}")
+    private String verifyUrl;
+
+    private final JavaMailSender mailSender;
 
     @Override
     @Async
-    public void send(String receiverEmail, String receiverFirstName, String token) {
-        String email = buildEmail(receiverFirstName, token);
+    public void sendVerificationMail(String receiverEmail, String receiverFirstName, String token) {
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: Just make the method more generic: send(email/s, subject, content)

**Review by todornikolov1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java (hunk 17 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java:33]: Yes, on row 64

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 18 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: The requested changes were made. Also a new FileServiceImpl was created, because the method for reading files is also not related to the emailService.

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 19 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: Method arguments were updated.

File: src/main/java/com/volasoftware/tinder/constants/Constants.java [hunk 20]
```diff
@@ -0,0 +1,6 @@
...
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/constants/Constants.java]: Name of classes in Java are usually singular.

Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/constants/Constants.java]: Actually, do we need this class after the constant class separation?

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 21]
```diff
@@ -35,12 +29,16 @@ public class EmailServiceImpl implements EmailService {
 
     private final JavaMailSender mailSender;
 
+    private final FileService fileService;
+
     @Override
     @Async
-    public void sendVerificationMail(String receiverEmail, String receiverFirstName, String token) {
-        String email = buildVerificationMail(receiverFirstName, token);
+    public void sendVerificationMail(String receiver, String token) {
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: this should be in user or auth service. EmailService class should send emails only

**Review by todornikolov1 — CHANGES_REQUESTED** [review]

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 19 above)
Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: I left new comment about that

File: src/main/java/com/volasoftware/tinder/constants/Constants.java (hunk 20 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/constants/Constants.java]: File was removed.

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java [hunk 22]
```diff
@@ -41,74 +37,4 @@ public void send(String receiverEmail, String receiverFirstName, String token) {
             throw new IllegalStateException("failed to send email");
```
Comment (Tsvetoslav-Garkov) [inline src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: This hardcoded message can be extracted as constant as well.

File: src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java (hunk 21 above)
Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/EmailServiceImpl.java]: method was moved to the AuthService class

File: src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java [hunk 23]
```diff
@@ -1,43 +1,65 @@
...
+    }
+
+    @Override
+    public Account verifyToken(String token) {
+
+        VerificationToken verificationToken = verificationTokenRepository.findByToken(token)
+                .orElseThrow(() -> new VerificationTokenExpiredException(SecurityConstant.TOKEN_EXPIRED));
 
-        emailService.send(account.getEmail(), account.getFirstName(), uuidToken);
+        isEmailAlreadyVerified(verificationToken);
```
Comment (Tsvetoslav-Garkov) [inline src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java:44]: One idea from me, not saying that must be implemented but for example say that the verification token expires and the user still has not validated his/her email, you want to clear this information by using some kind of schedular, because its obsolete. You can make sure user email has been validated by checking his profile validated flag, and not by checking the token get verified date is null. Because you can make the schedular clear also the tokens for which users have been verified, and not keeping unecessary data in the database. Think about it :)

Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java:44]: What you are suggesting is true, but in the requirements for that specific task, it is not specified whether the token must be deleted after it is no longer valid. 
Lets say thats the case, so after the token is deleted, even the user makes the call on the activation link, the token will not be found, therefore VerificationTokenExpiredException.class will be thrown and the user will ask for new verification token anyway (which task is in my next Sprint). 
Thank you for the information :)

Comment (Tsvetoslav-Garkov) [reply src/main/java/com/volasoftware/tinder/services/implementations/VerificationTokenServiceImpl.java:44]: The token is deleted only when it is expired or the user had verified the account by using valid one. If the user uses expired token or clicks again the link in the verification email we can just say that the token is invalid or has been used already. That was my idea.

File: src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java (hunk 17 above)
Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java:33]: Are you sure? I'm looking 64row, it says: 
`Account account = accountMapper.registerRequestToAccount(registerRequest); `
I don't see any usage of emailService there.

Comment (todornikolov1) [reply src/main/java/com/volasoftware/tinder/services/implementations/AuthenticationServiceImpl.java:33]: Oo, it is on  98 line now, as new code added

File: src/main/java/com/volasoftware/tinder/constants/MailConstant.java [hunk 24]
```diff
@@ -0,0 +1,11 @@
+package com.volasoftware.tinder.constants;
+
+public class MailConstant {
```
Comment (todornikolov1) [inline src/main/java/com/volasoftware/tinder/constants/MailConstant.java:3]: Are the constants real for mail?

Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/constants/MailConstant.java:3]: Without more requirements and further context about it, I think they are. What exactly do you mean?

Comment (tsenovi) [reply src/main/java/com/volasoftware/tinder/constants/MailConstant.java:3]: https://github.com/tsenovi/spring-tinder-application/pull/16/commits/4632b3329a2080234175c7b0c4f262cf9dadc9b7

**Review by todornikolov1 — APPROVED** [review]
